py/server/deephaven/_table_reader.py

```diff
@@ -0,0 +1,86 @@
+
+import jpy
+import numpy as np
+from deephaven import update_graph
+
+from deephaven.column import Column
+from deephaven.jcompat import to_sequence
+from deephaven.numpy import _column_to_numpy_array
+from deephaven.table import Table
+
+_JTableUpdateDataReader = jpy.get_type("io.deephaven.integrations.python.PythonTableDataReader")
+
+def _col_defs(table: Table, cols: Union[str, Sequence[str]]) -> Sequence[Column]:
+    if not cols:
+        col_defs = table.columns
+    else:
+        cols = to_sequence(cols)
+        col_defs = [col for col in table.columns if col.name in cols]
+
+    return col_defs
+
+def _table_reader_rows(table: Table, cols: Optional[Union[str, Sequence[str]]]) -> Generator[Dict[str, Any], None, None]:
+    """ A generator that reads one row at a time from a table into a dictionary. The dictionary is a map of column names
+    to scalar values of the column data type.
+
+    Args:
+        table (Table): The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+
+    Returns:
+        A generator that yields a dictionary of column names to a value.
+    """
+    col_defs = _col_defs(table, cols)
+
+    for chunk_dict in  _table_reader_chunks(table, cols=cols, row_set=table.j_table.getRowSet(), chunk_size=4096,
+                                            prev=False, to_numpy=False):
+        chunk_size = len(chunk_dict[col_defs[0].name])
+        for i in range(chunk_size):
+            col_dict = {}
+            for col_def in col_defs:
+                col_dict[col_def.name] = chunk_dict[col_def.name][i]
+            yield col_dict
+
+def _table_reader_chunks(table: Table, *, cols: Optional[Union[str, Sequence[str]]], row_set: jpy.JType, chunk_size: Optional[int],
+                         prev: bool = False, to_numpy: bool = True) -> Generator[Dict[str, np.ndarray], None, None]:
+    """ A generator that reads the chunks of rows over the given row set of a table into a dictionary. The dictionary is
+    a map of column names to numpy arrays.
+
+    Args:
+        table (Table):  The table to read.
+        cols (Optional[Union[str, Sequence[str]]]): The columns to read. If None, all columns are read.
+        row_set (jpy.JType): The row set to read.
+        chunk_size (Optional[int]): The number of rows to read at a time. If None, all rows in the row set are read.
+        prev (bool): If True, read the previous values. Default is False.
+        to_numpy (bool): If True, convert the column data to numpy arrays. Default is True.
+
+    Returns:
+        A generator that yields a dictionary of column names to numpy arrays.
+    """
+    col_defs = _col_defs(table, cols)
+
+    row_sequence_iterator = row_set.getRowSequenceIterator()
+    col_sources = [table.j_table.getColumnSource(col_def.name) for col_def in col_defs]
+    chunk_size = row_set.size() if not chunk_size else chunk_size
```

Comment:
I think I gave this feedback previously. We should not default to the entire table. If users want that, there are tools for taking a full snapshot.

Reply:
Not the whole table but the whole supplied row set, which makes sense when processing a TableUpdate.

Reply:
A row set could be the whole table; regardless, we should be setting a fixed size here when the user hasn't specified.

Reply:
I'm not opposed to the fixed size idea.  If we are really pushing this as a chunk iterator, it makes more sense to have a fixed size, and it is easier to explain.

Reply:
The whole chunk interface is to present "small chunks" to consumers. The idea being that consuming "small chunks" amortizes various costs associated that would otherwise be associated with non-chunky lookups, while also not blowing up memory.

If the caller doesn't specify a chunk size, the implementation should choose some reasonably "small" chunk size value. Typically, we use 1024, 2048, or 4096.

For example, in a 1B row parquet table, we _don't_ want to use a chunk size of 1B, as that would force everything into memory, largely defeating the advantages of consuming things in chunks.